### ISSUE TITLE
Accessing .Which if the constraint fails causes the rest of the code in AssertionScope to be skipped (immediate throw)

### ISSUE THREAD

**Issue by rbeurskens** [fluentassertions/fluentassertions#2664]
### Description

If I want to check for multiple conditions that I all want to report, and one happens to be a check on the existence of an item in a collection (and using `.Which` for additional checking in the item itself if it does exist), I do not expect `.Which` to throw if the Which Constraint<,> failed, but store the failure in the `AssertionScope` and ignore the constraint on `.Which` instead.

### Reproduction Steps

```csharp
// given similar API:
public class MyDataReader : IDataReader
{
    IEnumerable<IField> Fields { get; set;}
    IField GetField(string name) => Fields.SingleOrDefault(f=>f.ColumnName == name);
    // (...)
 }
public class Field : IField
{
    Field (string name, object value)
    {
        ColumnName = name;
        DataValue = value;
    }
    string ColumnName { get; }
    object DataValue { get; }
}

// Arrange
// values are both incorrect (swapped), column-/fieldname "Name" is different from expected "LastName"
var myRecord = new Field[] { new("FirstName", "Doe"), new("Name", "John") };
var dr = new MyDataReader { Fields = myRecord };
string input = "MyString";

// Act
bool result = dr.Read();

// Assert
result.Should().BeTrue();
using (new AssertionScope)
{
    // Expected "LastName", but actual is "Name"
    dr.Fields.Should().ContainSingle(field=>field.ColumnName=="LastName").Which.DataValue.Should().Be("Doe");
    // Because .Which above throws, the below code is not executed, but its results are expected to be included in the output.  
    dr.Fields.Should().ContainSingle(field=>field.ColumnName=="FirstName").Which.DataValue.Should().Be("John"); // <== Mismatches found here are not in the output
}
```

### Expected behavior

```
Expected dr.Fields to contain a single item matching (field.ColumnName == "LastName"), but no such item was found.
Expected dr.Fields to be "John" with a length of 4, but "Doe" has a length of 3, differs near "Doe" (index 0).
```

`// actual, the subject description for the check on .Which ('dr.Field') could also be improved ('dr.Fields.Item.DataValue'(?))`, but I think there is already another issue for that. (related to #1502 and/or #2253 ?)

### Actual behavior

```
Expected dr.Fields to contain a single item matching (field.ColumnName == "LastName"), but no such item was found.
```

### Known Workarounds

```C#
using (new AssertionScope)
{
    // Expected "LastName", but actual is "Name"
    using (var inner = new AssertionScope()) // isolated inner scope (in case there is an assertion added above that resulted in a failure)
        if (dr.Fields.Should().ContainSingle(field=>field.ColumnName=="LastName") is var constraint && !inner.HasFailures())
            constraint.Which.DataValue.Should().Be("Doe");
     // Because .Which above is now only accessed if there is no failure, it will not throw, so the below code is executed, and results are included in the output.
     // Has no effect here now, but there may be assertions added below in the future
     using (var inner = new AssertionScope()) // isolated inner scope (so we now hasFailure would always be caused by this assertion)
        if (dr.Fields.Should().ContainSingle(field=>field.ColumnName=="FirstName") is var constraint && !inner.HasFailures())
            constraint.Which.DataValue.Should().Be("John"); // <== Mismatches found here are now in the output
}
```
or
```C#
using (var scope = new AssertionScope)
{
    // using extension (A)
    if (scope.IsSuccessful(() => dr.Fields.Should().ContainSingle(field => field.ColumnName == "LastName")
        // dr.GetField("LastName").DataValue.Should().Be("Doe");
        dr.Fields.Single(field => field.ColumnName == "LastName").DataValue.Should().Be("Doe");
    // (...)
    // or (B)
    scope.IfSuccessful(() => dr.Fields.Should().ContainSingle(field => field.ColumnName == "LastName")
        field => field.DataValue.Should().Be("Doe"));
        // _ => dr.Fields.Single(field => field.ColumnName == "LastName").DataValue.Should().Be("Doe"); // would work too and adds some more context to the output.
    // (...)
}

public static class FluentAssertionExtensions
{
    // extending on AssertionScope to reduce chance to have them apply in code where they are not relevant

    // alternative A
    public static bool IsSuccessful(this IAssertionScope _ , Action evaluateConstraint)
    {
        using var scope = new AssertionScope();
        evaluateConstraint();
        return !scope.HasFailures();
    }
    // alternative B
    public static void IfSuccessful<TParent, T>(this IAssertionScope _ , Func<AndWhichConstraint<TParent,T>> constraintFunc, Action<T> evaluateItem)
    {
        using var scope = new AssertionScope();
        var constraint = constraintFunc();
        if (!scope.HasFailures())
             evaluateItem(constraint.Which);
    }
    // alternative C (quick and dirty - use of exception for control flow)
    public static void IfSuccessful<TParent, T>(this AndWhichConstraint<TParent, T> constraint, Action<T> code)
    {
        T subject;
        try { subject = constraint.Which!; }
        catch (Exception) { return; }
        code(subject);
    }
}
```

### Are you willing to help with a pull-request?

No

**Comment by dennisdoomen** [comment]
This is a scenario we're trying to resolve with #2253.

**Comment by dennisdoomen** [comment]
Even after the introduction of the new `AssertionChain` on the `develop` branch, your particular example will never work. Here's why:

Let's start with the first statement.

    dr.Fields.Should().ContainSingle(field => field.ColumnName == "LastName")

It will fail the assertion, report the failure in the surrounding `AssertionScope`, and return an instance of `AndWhichConstraint` that contains an empty collection of `Field`. As soon as you invoke the `Which` property, it'll find a single item (which doesn't exist) and throw an exception. 

No imagine we fix the `Which` to return `null` in that case, you're example will throw an NRE because you're getting the `DataValue` property from a `null` object. Now, we _can_ change your example to use `Which?.DataValue`, which will ensure that the `Should().Be()` is never executed. 

But because how the first `ContainSingle` tries to pass the `AssertionChain` to the second `Be` assertion, and that one is never invoked (because of the `Which?` nullable invocation), the next call to `ContainSingle` will pick up the thread-local `AssertionChain`. Since the original assertion failed, no other assertion will be executed. 

In other words, there's no way we can fix this particular scenario.

**Comment by rbeurskens** [comment]
I think I understand why it won't work as long as `.Which` is a property that returns a direct reference to the resulting subject of a previous assertion. I have the above workaround for this kind of scenario to check the current (inner) `AssertionScope` before accessing it for further assertion. Using nullibility on the `Which` property for control flow would not be a good idea anyway imo. (what if we want to find a `null` item in a collection it is of a value type? ;) ) For that I think we would need either the explicit `if` before accessing `Which` or an API allowing to provide lazy chained assertions such as `TContstraint AndWhichContsraint<TParent,T>.Which<TContstraint>(Func<T,TContstraint> chainedAssertion)` where the provided delegate will only be invoked if the previous assertion was successful,  which is what I try to do using the extension methods earlier.
